### ISSUE TITLE
Wrongly computed PCH header dir

### ISSUE THREAD

**Issue by Goraf** [Caphyon/clang-power-tools#209]
For directory structure
```
PROJECT_DIR/
    src/
        stdafx.cxx
        stdafx.h
```
and project file importing those properties
```
  <ItemGroup>
    <ClCompile Include="src\stdafx.cxx">
      <PrecompiledHeader>Create</PrecompiledHeader>
    </ClCompile>
    <ClCompile Include="src\*.cpp" />
    <ClCompile Include="src\*.c">
      <PrecompiledHeader>NotUsing</PrecompiledHeader>
    </ClCompile>
    <ClInclude Include="src\*.hpp" />
    <ClInclude Include="src\*.h" />
  </ItemGroup>
```
running `Tidy Code` with `Clang Power Tool 2.5.1` and `VS2017` outputs
```
VERBOSE: PCH cpp name: src\stdafx.cxx
VERBOSE: PCH header name: stdafx.h
Join-Path : Cannot find path 'Disc:\...\PROJECT_DIR\stdafx.h' because it does not exist.
At C:\Users\...\VisualStudio\15.0_7afaccbd\Extensions\4zo3oigp.tpb\clang-build.ps1:524 char:25
+ ... ]] $paths = Join-Path -Path "$base" -ChildPath "$child" -Resolve -Err ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Disc:\...\PROJECT_DIR\stdafx.h:String) [Join-Path], ItemNotFoundExce 
   ption
    + FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.JoinPathCommand
```
which is wrong because file is at 'Disc:\...\PROJECT_DIR\src\stdafx.h'

Any workarounds? `<PrecompiledHeaderFile>` property doesn't accept path, only filename

**Comment by hero101111** [comment]
Hello,

Please attach a sample project we can test.

**Comment by Goraf** [comment]
Hi,
it's an opensource project with code here (https://github.com/miranda-ng/miranda-ng) but the code base is kinda big. Do you want me to try to create/reproduce some minimal example?

**Comment by hero101111** [comment]
That would be great, yes.

**Comment by Goraf** [comment]
Looks like it is very easy to reproduce so I'll just describe it.

1. Create new sample project (`Visual C++` -> `Windows Desktop` -> `Dynamic-Link Library (DLL)`)
1. Run `Clang Tidy`. It will process files (kinda)
1. Add subdirectory `src` to project directory
1. Rename `stdafx.cpp` file to `stdafx.cxx`
1. Move all source/header files (*.cpp, *.cxx, *.h) to `src` subdirectory
1. Edit project file and replace
```
  <ItemGroup>
    <ClInclude Include="stdafx.h" />
    <ClInclude Include="targetver.h" />
  </ItemGroup>
  <ItemGroup>
    <ClCompile Include="Dll2.cpp" />
    <ClCompile Include="dllmain.cpp" />
    <ClCompile Include="stdafx.cpp">
      <PrecompiledHeader Condition="'$(Configuration)|$(Platform)'=='Debug|Win32'">Create</PrecompiledHeader>
      <PrecompiledHeader Condition="'$(Configuration)|$(Platform)'=='Debug|x64'">Create</PrecompiledHeader>
      <PrecompiledHeader Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">Create</PrecompiledHeader>
      <PrecompiledHeader Condition="'$(Configuration)|$(Platform)'=='Release|x64'">Create</PrecompiledHeader>
    </ClCompile>
  </ItemGroup>
```
with
```
  <ItemGroup>
    <ClCompile Include="src\stdafx.cxx">
      <PrecompiledHeader>Create</PrecompiledHeader>
    </ClCompile>
    <ClCompile Include="src\*.cpp" />
    <ClCompile Include="src\*.c">
      <PrecompiledHeader>NotUsing</PrecompiledHeader>
    </ClCompile>
    <ClInclude Include="src\*.hpp" />
    <ClInclude Include="src\*.h" />
  </ItemGroup>
```
- Reload project
- Run `Clang Tidy` and you will end up with mentioned error
[Dll2.zip](https://github.com/Caphyon/clang-power-tools/files/1652046/Dll2.zip)

**Comment by hero101111** [comment]
Hello,

Thank you for the sample project; it's been very helpful while investigating the issue.
We found that the root cause were the wildcard includes for the header/source files: Clang Power Tools was not able to handle them.

We've just pushed a fix that fixes the issue. 
An updated version will be released shortly.

It would be of great help to us if you could manually test our fix and confirm that everything's working.
See the instructions below.

Regards,
Gabriel

---

Please note that this fix is currently live only for the github repository.
The latest Visual Studio extension (`2.5.1`) does not contain the fix. It will be included in the next release.

Until then, as a workaround, you can manually replace `clang-build.ps1` in the VS extension directory with the latest from github. To locate the extension directory search for `clang-build.ps1` in `%LOCALAPPDATA%`

Don't forget to restart Visual Studio after replacing the script file.

**Comment by Goraf** [comment]
So quick test shows that it works now in most cases. But there is one project with which it fails in the same place and with the same error as before. I failed to reproduce it with some simple example for now. I don't have a clue what could be the reason, as it imports and uses almost the same general properties as rest of the projects. Will try to figure it out.

**Comment by Goraf** [comment]
Scenario which fails:
All projects have the same directories structure 
```
PROJECT_DIR/
    project_file
    src/
        stdafx.cxx
        stdafx.h
        <source_files>
```
so in a project file they import mentioned before properties 
```
  <ItemGroup>
    <ClCompile Include="src\stdafx.cxx">
      <PrecompiledHeader>Create</PrecompiledHeader>
    </ClCompile>
    <ClCompile Include="src\*.cpp" />
    <ClCompile Include="src\*.c">
      <PrecompiledHeader>NotUsing</PrecompiledHeader>
    </ClCompile>
    <ClInclude Include="src\*.hpp" />
    <ClInclude Include="src\*.h" />
  </ItemGroup>
```
from a common properties sheet with an entry like
`  <Import Project="common.props" />`
And to that point fix to the script works.

Now, lets say some project has its own unique, internal library in its sources, so this project's structure looks like  
```
PROJECT_DIR/
    project_file
    src/
        stdafx.cxx
        stdafx.h
        <source_files>
        library/
            <library_source_files>
```
so it also needs additional include entries in its project file after imported common includes. Like that
```
  <Import Project="common.props" />
  <ItemGroup>
    <ClCompile Include="src\library\*.cpp">
      <PrecompiledHeader>NotUsing</PrecompiledHeader>
    </ClCompile>
    <ClInclude Include="src\library\*.h" />
  </ItemGroup>
```
And in such case script fails.

[sample project.zip](https://github.com/Caphyon/clang-power-tools/files/1656181/Dll2.zip)

**Comment by hero101111** [comment]
Hello,

We've investigated your latest sample project and found that there are issues when project source files / headers are specified in property sheets. At the moment our project data loader is unable to load them from propsheets. We are working on a project data loader fix regarding this.

We'll inform you when the fix will be pushed. Stay posted.

Regards,
Gabriel

**Comment by cpp-red-lion** [comment]
This issue was _partially_ resolved in version `2.6`

The _extended_ related case, when using VS *property-sheets*, will be fixed in a following version and moved to issue: #219
